packages/buidler-core/src/internal/sentry/anonymizer.ts

```diff
@@ -0,0 +1,152 @@
+import * as path from "path";
+
+const ANONYMIZED_FILE = "<user-file>";
+
+export class Anonymizer {
+  constructor(private _configPath?: string) {}
+
+  /**
+   * Given a sentry serialized exception
+   * (https://develop.sentry.dev/sdk/event-payloads/exception/), return an
+   * anonymized version of the event.
+   */
+  public anonymize(event: any): either.Either<string, Event> {
+    if (event === null || event === undefined) {
+      return either.left("event is null or undefined");
+    }
+    if (typeof event !== "object") {
+      return either.left("event is not an object");
+    }
+
+    const result: Event = {
+      event_id: event.event_id,
+      platform: event.platform,
+      timestamp: event.timestamp,
+    };
+
+    if (event.exception !== undefined && event.exception.values !== undefined) {
+      const anonymizededExceptions = this._anonymizeExceptions(
+        event.exception.values
+      );
+      result.exception = {
+        values: anonymizededExceptions,
+      };
+    }
+
+    return either.right(result);
+  }
+
+  private _anonymizeExceptions(exceptions: Exception[]): Exception[] {
+    return exceptions.map((exception) => this._anonymizeException(exception));
+  }
+
+  private _anonymizeException(value: Exception): Exception {
+    const result: Exception = {
+      type: value.type,
+      value: value.value,
+    };
+
+    if (value.stacktrace !== undefined) {
+      result.stacktrace = this._anonymizeStacktrace(value.stacktrace);
+    }
+
+    return result;
+  }
+
+  private _anonymizeStacktrace(stacktrace: Stacktrace): Stacktrace {
+    if (stacktrace.frames !== undefined) {
+      const anonymizededFrames = this._anonymizeFrames(stacktrace.frames);
+      return {
+        frames: anonymizededFrames,
+      };
+    }
+
+    return {};
+  }
+
+  private _anonymizeFrames(frames: StackFrame[]): StackFrame[] {
+    return frames.map((frame) => this._anonymizeFrame(frame));
+  }
+
+  private _anonymizeFrame(frame: StackFrame): StackFrame {
+    const result: StackFrame = {
+      lineno: frame.lineno,
+      colno: frame.colno,
+      function: frame.function,
+    };
+
+    let anonymizeContent = true;
+    if (frame.filename !== undefined) {
+      const anonymizationResult = this._anonymizeFilename(frame.filename);
+      result.filename = anonymizationResult.anonymizedFilename;
+      anonymizeContent = anonymizationResult.anonymizeContent;
+    }
+
+    if (!anonymizeContent) {
+      result.context_line = frame.context_line;
+      result.pre_context = frame.pre_context;
+      result.post_context = frame.post_context;
+      result.vars = frame.vars;
+    }
+
+    return result;
+  }
+
+  /**
+   * Return the anonymized filename and a boolean indicating if the content of
+   * the file should be anonymized
+   *
+   */
+  private _anonymizeFilename(
+    filename: string
+  ): { anonymizedFilename: string; anonymizeContent: boolean } {
+    if (filename === this._configPath) {
```

Comment:
This function is very complex and could use some tests.

If `packageJsonPath` is made a parameter instead of being computed, then this function becomes pure (or, well, pure enough) and testing it would be very straightforward. The downside is that it makes things weird for the callee (it has to call `findup` for all files) and we have to make this method public (which I don't mind).